internal/datastore/crdb/pool/pool.go

```diff
@@ -0,0 +1,393 @@
+	if disabled, ok := ctx.Value(CtxDisableRetries).(bool); ok && disabled {
+		maxRetries = 0
+	}
+
+	for retries = uint32(0); retries <= maxRetries; retries++ {
+		err = wrapRetryableError(ctx, fn(conn))
+		if err == nil {
+			conn.Release()
+			if retries > 0 {
+				log.Ctx(ctx).Info().Uint32("retries", retries).Msg("resettable database error succeeded after retry")
+			}
+			return nil
+		}
+
+		var (
+			resettable *ResettableError
+			retryable  *RetryableError
+		)
+		if errors.As(err, &resettable) {
+			log.Ctx(ctx).Info().Err(err).Uint32("retries", retries).Msg("resettable error")
+
+			id := p.safeGC(ctx, conn)
+			conn.Release()
+
+			// After a resettable error, mark the node as unhealthy
+			// TODO: configurable error count / circuit-breaker
+			if id > 0 {
+				p.healthTracker.SetNodeHealth(id, false)
+			}
+
+			sleepOnErr(ctx, err, retries)
+
+			conn, err = p.acquireFromDifferentNode(ctx, id)
+			if err != nil {
+				return fmt.Errorf("error acquiring connection from pool after retry %d: %w", retries, err)
+			}
+			continue
+		}
+		if errors.As(err, &retryable) {
+			log.Ctx(ctx).Info().Err(err).Uint32("retries", retries).Msg("retryable error")
+			sleepOnErr(ctx, err, retries)
+			continue
+		}
+		conn.Release()
+		// error is not resettable or retryable
+		log.Ctx(ctx).Warn().Err(err).Uint32("retries", retries).Msg("error is not resettable or retryable")
+		return err
+	}
+	return &MaxRetryError{MaxRetries: p.maxRetries, LastErr: err}
+}
+
+// safeGC recovers if the underlying puddle's connection is nil
+// It returns a nodeID of zero if so, since we lost access to the connection object
+// to compute the id from.
+func (p *RetryPool) safeGC(ctx context.Context, conn *pgxpool.Conn) uint32 {
+	var id uint32
+	defer func() {
+		if r := recover(); r != nil {
+			log.Ctx(ctx).Warn().Msg("recovered from a missing pgx.Conn on a pgxpool.Conn")
+		}
+	}()
+	id = p.GC(conn.Conn())
+	return id
+}
+
+// GC marks a connection for destruction on the next Acquire.
+// BeforeAcquire can signal to the pool to close the connection and clean up
+// the reference in the pool at the same time, so we lazily GC connections
+// instead of closing the connection directly.
+// It returns the node id of the connection that was just closed.
+func (p *RetryPool) GC(conn *pgx.Conn) uint32 {
+	p.Lock()
+	defer p.Unlock()
+	p.gc[conn] = struct{}{}
+	node := p.nodeForConn[conn]
+	delete(p.nodeForConn, conn)
+	return node
+}
+
+func sleepOnErr(ctx context.Context, err error, retries uint32) {
+	after := retry.BackoffExponentialWithJitter(100*time.Millisecond, 0.5)(uint(retries + 1)) // add one so we always wait at least a little bit
+	log.Ctx(ctx).Warn().Err(err).Dur("after", after).Msg("retrying on database error")
+	time.Sleep(after)
+}
+
+func (p *RetryPool) acquireFromDifferentNode(ctx context.Context, nodeID uint32) (*pgxpool.Conn, error) {
+	log.Ctx(ctx).Info().Uint32("node_id", nodeID).Msg("acquiring a connection from a different node")
+	for {
```

Comment:
it could, though we are also in parallel attempting to detect bad nodes and replace all of their connections. Not sure what else to do here though, it doesn't make sense to acquire a connection to a draining node.

Reply:
yeah I don't have an alternative either, and I guess in the grand schema of things, at least it's only acquiring one connection at a time.